File: led-client/handshake.py
# handshake.py
from request import fetch_data
from getMac import get_mac_address
from models import Device

def handshake(connection, relative_endpoint) -> bool | Device:
    """Sends a request to the server to register the device"""
    
    mac = get_mac_address()
    print(f'got mac: {mac}')
    
    handshake_data = {
        'mac': mac,
        'type': 'LedStrip',
        'ip': connection.ip,
    }
    
    try:
        print(f'attempting handshake with at "{relative_endpoint}"')
        res_json = fetch_data(relative_endpoint, method="POST", data={"data": handshake_data})
        
        if res_json is None:
            print('handshake failed')
            return False
        
        data = res_json.get('data')
        if error := res_json.get('error'):
            print(f'received error: {error}')
            return False
        
        device = Device(data)
        print(f'received response:\n', res_json)
        device.write()
        return device
    
    except Exception as e:
        print('error during handshake:', e)
        return False

File: pico-client-c/handshake.c
#include "handshake.h"
#include "getMac.h"
#include "models.h"
#include "request.h"
#include "networkConnection.h"
#include "cJSON.h"
#include <stdio.h>
#include <stdlib.h>

bool handshake(struct NetworkConnection *connection, const char *relative_endpoint, struct Device *device) {
    char *mac = get_mac_address();
    printf("got mac: %s\n", mac);

    cJSON *handshake_data = cJSON_CreateObject();
    cJSON_AddStringToObject(handshake_data, "mac", mac);
    cJSON_AddStringToObject(handshake_data, "type", "LedStrip");
    cJSON_AddStringToObject(handshake_data, "ip", connection->ip);

    cJSON *post_data = cJSON_CreateObject();
    cJSON_AddItemToObject(post_data, "data", handshake_data);

    char *body = cJSON_PrintUnformatted(post_data);
    cJSON_Delete(post_data);

    printf("attempting handshake with at \"%s\"\n", relative_endpoint);
    cJSON *res_json = fetch_data(relative_endpoint, "POST", body);
    free(body);

    if (res_json == NULL) {
        printf("handshake failed\n");
        return false;
    }

    cJSON *error = cJSON_GetObjectItem(res_json, "error");
    if (error) {
        printf("received error: %s\n", error->valuestring);
        cJSON_Delete(res_json);
        return false;
    }

    cJSON *data = cJSON_GetObjectItem(res_json, "data");
    device_init(device, data);
    printf("received response:\n%s\n", cJSON_Print(res_json));
    device_write(device);
    cJSON_Delete(res_json);
    return true;
}
